hummingbot/connector/gateway/clob_spot/data_sources/gateway_clob_api_data_source_base.py

```diff
@@ -0,0 +1,130 @@
+import logging
+from abc import ABC, abstractmethod
+from decimal import Decimal
+from enum import Enum
+from typing import Any, Callable, Dict, List, Mapping, Optional, Tuple
+
+from bidict import bidict
+
+from hummingbot.connector.gateway.gateway_in_flight_order import GatewayInFlightOrder
+from hummingbot.connector.gateway.gateway_order_tracker import GatewayOrderTracker
+from hummingbot.connector.trading_rule import TradingRule
+from hummingbot.core.data_type.in_flight_order import InFlightOrder, OrderUpdate, TradeUpdate
+from hummingbot.core.data_type.order_book_message import OrderBookMessage
+from hummingbot.core.data_type.trade_fee import MakerTakerExchangeFeeRates
+from hummingbot.core.event.event_forwarder import EventForwarder
+from hummingbot.core.event.event_listener import EventListener
+from hummingbot.core.event.events import AccountEvent, MarketEvent, OrderBookDataSourceEvent
+from hummingbot.core.network_iterator import NetworkStatus
+from hummingbot.core.pubsub import HummingbotLogger, PubSub
+
+
+class GatewayCLOBAPIDataSourceBase(ABC):
+    _logger: Optional[HummingbotLogger] = None
+
+    @classmethod
+    def logger(cls) -> HummingbotLogger:
+        if cls._logger is None:
+            cls._logger = logging.getLogger(HummingbotLogger.logger_name_for_class(cls))
+        return cls._logger
+
+    def __init__(self):
+        self._publisher = PubSub()
+        self._forwarders_map: Dict[Tuple[Enum, Callable], EventForwarder] = {}
+        self._gateway_order_tracker: Optional[GatewayOrderTracker] = None
+
+    @property
+    def gateway_order_tracker(self):
+        return self._gateway_order_tracker
+
+    @gateway_order_tracker.setter
+    def gateway_order_tracker(self, tracker: GatewayOrderTracker):
+        if self._gateway_order_tracker is not None:
+            raise RuntimeError("Attempted to re-assign the order tracker.")
+        self._gateway_order_tracker = tracker
+
+    @staticmethod
+    def supported_stream_events() -> List[Enum]:
+        return [
+            MarketEvent.TradeUpdate,
+            MarketEvent.OrderUpdate,
+            AccountEvent.BalanceEvent,
+            OrderBookDataSourceEvent.TRADE_EVENT,
+            OrderBookDataSourceEvent.DIFF_EVENT,
+            OrderBookDataSourceEvent.SNAPSHOT_EVENT,
+        ]
+
+    def add_forwarder(self, event_tag: Enum, receiver: Callable):
+        event_forwarder = EventForwarder(to_function=receiver)
+        self.add_listener(event_tag=event_tag, listener=event_forwarder)
+        self._forwarders_map[(event_tag, receiver)] = event_forwarder
```

Comment:
This clarifies for me some of your previous comments. To be honest, I hadn't taken into consideration the automatic forwarder "garbage collection" by the PubSub when designing this. My focus was to hide the forwarder management from the user and have the user only worry about the event tag and the callback. But I see your point, we are trading fault-tolerance for convenience here... Maybe we can discuss this over a call. I believe that, if the user fails to unsubscribe, then this can signal a cleanup routine that failed to run, which is either ways something we should address...